src/shared/vscode/window.ts

```diff
@@ -74,16 +103,55 @@ class DefaultWindow implements Window {
         return vscode.window.showInputBox(options, token)
     }
 
-    public showInformationMessage(message: string, ...items: string[]): Thenable<string | undefined> {
-        return vscode.window.showInformationMessage(message, ...items)
+    public showInformationMessage(message: string, ...items: string[]): Thenable<string | undefined>
+    public showInformationMessage(
+        message: string,
+        options: vscode.MessageOptions,
+        ...items: string[]
+    ): Thenable<string | undefined>
+    public showInformationMessage<T extends vscode.MessageItem>(message: string, ...items: T[]): Thenable<T | undefined>
+    public showInformationMessage<T extends vscode.MessageItem>(
+        message: string,
+        options: vscode.MessageOptions,
+        ...items: T[]
+    ): Thenable<T | undefined>
```

Comment:
do these signatures not already inherit from `implements Window`?

Reply:
It works without them, but then typescript won't restrict the allowable values to those in the overloads.

```typescript
interface Foo {
    bar(x: string): void
    bar(x: number): void
}

class FooImpl implements Foo {
    bar(x: any): void {
        console.log(x)
    }
}

new FooImpl().bar({}) // Just wanted strings...
```

[Playground Link](https://www.typescriptlang.org/play/index.html?ssl=28&ssc=35&pln=2&pc=1#)

Reply:
one can use `as Foo` to get typescript checking:

    (new FooImpl() as Foo).bar({}) // error

The added boilerplate in this PR is not worth the gain imo.

Reply:
I would say it's worth it because it's documenting the implementation in that class. Also, with the solution above, you lose type safety.

```typescript
;({} as Foo).bar('blah') // No error, but {} isn't Foo. It's a trap!
```

Reply:
It is hundreds of lines of code to avoid a mistake that would only be caught if the codepath isn't tested.

Reply:
Checking again looks like `Default*` classes aren't exported and the users of the `Fake*` classes shouldn't invoke these APIs directly anyway, so that helps ease my concerns in this case. Will remove the ones in the impls.

In general though, this just seems like the price to pay for overloads / structural typing in TS.